server/src/main/java/io/crate/replication/logical/action/TransportDropSubscriptionAction.java

```diff
@@ -21,130 +21,283 @@
-                Metadata currentMetadata = currentState.metadata();
-                Metadata.Builder mdBuilder = Metadata.builder(currentMetadata);
-
-                SubscriptionsMetadata oldMetadata = (SubscriptionsMetadata) mdBuilder.getCustom(SubscriptionsMetadata.TYPE);
-                if (oldMetadata != null && oldMetadata.subscription().containsKey(request.name())) {
-
-                    // LogicalReplicationService will react to cluster change events and will stop the replication.
-                    SubscriptionsMetadata newMetadata = SubscriptionsMetadata.newInstance(oldMetadata);
-                    newMetadata.subscription().remove(request.name());
-                    assert !newMetadata.equals(oldMetadata) : "must not be equal to guarantee the cluster change action";
-                    mdBuilder.putCustom(SubscriptionsMetadata.TYPE, newMetadata);
-
-                    return ClusterState.builder(currentState).metadata(mdBuilder).build();
-                } else if (request.ifExists() == false) {
-                    throw new SubscriptionUnknownException(request.name());
-                }
+    protected void masterOperation(DropSubscriptionRequest request, ClusterState state, ActionListener<AcknowledgedResponse> listener) throws Exception {
+        List<IndexMetadata> subscriptionIndices = getSubscriptionIndices(request.name(), state.metadata().indices());
+        List<OpenCloseTable> openCloseTables = getSubscriptionTables(request.name(), schemas);
 
-                return currentState;
-            }
-        };
+        submitCloseSubscriptionsTablesTask(openCloseTables)
+            .thenCompose(res -> submitDropSubscriptionTask(request.name(), request.ifExists(), subscriptionIndices))
+            .thenCompose(res -> submitOpenSubscriptionsTablesTask(openCloseTables))
+            .whenComplete(
+                (ignore, err) -> {
+                    if (err == null) {
+                        listener.onResponse(new AcknowledgedResponse(true));
+                    } else {
+                        // Hint in logs with action to repeat is already shown on the corresponding step.
+                        listener.onFailure(Exceptions.toRuntimeException(err));
+                    }
+                }
+            );
     }
 
     @Override
     protected ClusterBlockException checkBlock(DropSubscriptionRequest request, ClusterState state) {
         return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_WRITE);
     }
+
+    /**
+     * Closes tables and consequently closes all shards of the index
+     * what, in turn, stops trackers and finally,
+     * stopping tracker removes retention lease on the remote cluster.
+     */
+    private CompletableFuture<Void> submitCloseSubscriptionsTablesTask(List<OpenCloseTable> openCloseTables) {
+        var future = new CompletableFuture<Void>();
+
+        clusterService.submitStateUpdateTask(
```

Comment:
`.isLocalNodeElectedMaster()` is checked only on second and third step - first step is already on master